src/DotNetty.Buffers/ByteBufferUtil.cs

```diff
@@ -569,5 +569,12 @@ static void AppendHexDumpRowPrefix(StringBuilder dump, int row, int rowStartInde
                 dump.Append('|');
             }
         }
+
+        public static string DecodeString(IByteBuffer src, int readerIndex, int len, Encoding encoding)
+        {
+            if (len == 0) return string.Empty;
+            
+            return encoding.GetString(src.Array.Slice(len));
```

Comment:
`HasArray` is actually an indicator, whether the byte buffer's content is available through `Array` property directly. So for `CompositeByteBuffer` with 2+ components it will be `false`. so really in case of `buf.HasArray == false` you'd need to read it out into array-backed buffer.